### ISSUE TITLE
Certain on-load GA events are being sent before the pageview

### ISSUE THREAD

**Issue by chrisblakley** [chrisblakley/Nebula#1415]
Traffic that sends server-side events without a landing page (typically bots) appear to actually be triggering other events. I want to make sure these are real visitors, and if not I'd like to come up with a way to filter out this traffic.

 - It's affecting  the New Users metric
 - I'm afraid it's affecting the performance timing metrics
 - I'm afraid it's affecting the load abandonment metrics

Here is an event report segmented to `(not set)` landing page. With this same segment, the `All Pages` report has no rows. One particular event that makes me think most of these are bots: `Login Error` > `ERROR: Cookies are blocked or not supported by your browser...`

![screen shot 2017-07-13 at 9 48 53 am](https://user-images.githubusercontent.com/6349678/28170147-725828a0-67b2-11e7-8379-32ed291a481c.png)

**Comment by chrisblakley** [comment]
I want to avoid a global filter for `(not set)` landing pages, so is there something else that we can do?

Nebula doesn't know if a user is a bot when their user agent is set to a normal browser. One thing we can check is if these are AJAX events (Eg: GA is blocked).

I also want to re-double-check that the server-side events are sending page data.

**Comment by chrisblakley** [comment]
Similar to these other issues:
 - https://github.com/chrisblakley/Nebula/issues/1070
 - https://github.com/chrisblakley/Nebula/issues/1389

**Comment by chrisblakley** [comment]
**I think the issue is that server-side events, even when using `dl` and `dt` are still an `event` hit type, so the landing page will always be `(not set)` because a `pageview` hit type wasn't yet sent.**

I think I'm going to have to look over all server-side events with this thought process...

**Comment by chrisblakley** [comment]
Internal search events are now only sending for non-bots and when the GA CID was not generated by Nebula.

**Comment by chrisblakley** [comment]
```js
jQuery('h1').on('click', function(){
	console.log('clicked h1');
	jQuery.ajax({
		type: "POST",
		url: nebula.site.ajax.url,
		data: {
			nonce: nebula.site.ajax.nonce,
			action: 'nebula_test_cookie',
		},
		success: function(response){
			console.log('ajax success');
			console.debug(response);
		},
		timeout: 60000
	});
});
```

```php
add_action('wp_ajax_nebula_test_cookie', 'nebula_test_cookie');
add_action('wp_ajax_nebula_test_cookie', 'nebula_test_cookie');
function nebula_test_cookie(){
	echo 'inside ajax...... ';
	var_dump( nebula()->ga_parse_cookie() );
}
```

Reads the correct cookie, so I think the AJAX events are fine as-is without needing to pass the CID as an extra parameter.

**Comment by chrisblakley** [comment]
Alright, going to try pushing that to a few sites and see if it improves this.

**Comment by chrisblakley** [comment]
Seeing a smaller amount, but still getting some... This time all on JS events (no server-side):

![screen shot 2017-07-16 at 7 30 33 pm](https://user-images.githubusercontent.com/6349678/28252248-4fdda7ca-6a5d-11e7-9e50-6b54dc5a8630.png)

**Comment by chrisblakley** [comment]
Definitely not enough data yet, but it's looking like the `(not set)` landing page events are JavaScript.

**Comment by chrisblakley** [comment]
Ok, here's my new thought!

Nebula checks when the GA tracker is available and triggers a flag immediately that tells the event tracker that it's ready. Could this also be allowing for events to be sent prior to the pageview if main.js is also ready at that point?

**Comment by chrisblakley** [comment]
Saw a few events of "Non-linked click attempt" and "Performance timing" still associated with `(not set)` landing pages...

**Comment by chrisblakley** [comment]
![oco](https://user-images.githubusercontent.com/6349678/28925576-3cc9e6ec-7833-11e7-9ac0-4ec60e7ff3e7.gif)

Performance timing and initial scroll are still appearing as `(not set)`.

Ok big idea this time- I'm not setting the `gaready` flag until the pageview callback, so now main.js should not know GA is ready until either the pageview has been sent *or* window load.

Will check back in a few more days.

**Comment by chrisblakley** [comment]
![oco](https://user-images.githubusercontent.com/6349678/28925576-3cc9e6ec-7833-11e7-9ac0-4ec60e7ff3e7.gif)

One more `(not set)` event appeared today for Scroll Depth...

I'm out of ideas at this point. Gonna give it a few more days.

**Comment by chrisblakley** [comment]
Ok this time I've disabled the `performanceMetrics()` function altogether. Even though it is called on window load...

**Comment by chrisblakley** [comment]
Had a bunch more appear today... This time a handful of `Non-Linked Click Attempt` and `Scroll Depth`. What I still don't understand is that these events were **all** sent after the pageview (as proven by my custom dimension). None of my event listeners are even active until after the pageview, so it might be something happening server side, in `analytics.php`, or some other hit type (exception or timing maybe?).

I might disable Autotrack next just to take it out of the equation.

**Comment by chrisblakley** [comment]
Ok, I can definitely confirm that **something** is being sent before the pageview. I made a new segment for sessions with a hit being sent `Before pageview`. Still need to make sure this isn't the pageview hit itself.

Even if it isn't, there isn't a way (that I can tell) to find explicitly which hit was the one with that dimension... Maybe in the User Explorer?

**Comment by chrisblakley** [comment]
I'm seeing a gap of 4+ hours in one case which triggered a new session, but no new pageview. Then an event was triggered which would cause a `(not set)` landing page... Could this be a common occurrence here?

**Comment by chrisblakley** [comment]
Moved the tracker part of this to a separate issue: https://github.com/chrisblakley/Nebula/issues/1455

I've made the other updates, so check back on those next week.

**Comment by chrisblakley** [comment]
Increasing session duration appears to fix the problem, but it isn't going to completely prevent this issue:

![screen shot 2017-08-13 at 11 31 52 am](https://user-images.githubusercontent.com/6349678/29251003-2229d988-801b-11e7-8de4-07249e0e7b5e.png)

Sometime this week I'm going to slowly start bringing events back.

Security Precautions have all been changed to the exception hit type.

**Comment by chrisblakley** [comment]
In addition to slowly bringing things back, I've also added a counter for session pages. Server-side events will not be sent unless this counter is at 2 or more.

**Comment by chrisblakley** [comment]
Here's the next enhancement: I added a Nebula Option that will listen for the last GA hit and if the next one is beyond the session timeout it will send a new pageview.

```
<?php if ( nebula()->get_option('ga_session_timeout_minutes') && intval(nebula()->get_option('ga_session_timeout_minutes')) >= 5 ): ?>
	if ( model.get('hitType') !== 'pageview' && typeof lastHit === 'object' ){
		var currentHit = new Date();
		if ( (currentHit-lastHit) > (<?php echo nebula()->get_option('ga_session_timeout_minutes'); ?>*60000) ){ //If after GA session timeout
			//@TODO "Nebula" 0: Set dimension for pageview or give channel or utm parameters so pageview doesn't show up as just 'direct'?
			ga('send', 'pageview');
		}
	}
	lastHit = new Date(); //Update the last GA hit time
<?php endif; ?>
```

**Comment by chrisblakley** [comment]
Ok, I'm officially calling this one done. Even though ultimately the **correct** answer was to increase GA Session Timeout, there were multiple major breakthroughs for Nebula here.
